src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,186 @@
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw  new IndexOutOfBoundsException("Index is out of bound");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            if (head != null) {
+                head.prev = newNode;
+            }
+            head = newNode;
+            if (tail == null) {
+                tail = newNode;
+            }
+            size++;
+            return;
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        Node<T> newNode = new Node<>(current.prev, value, current);
+        if (current.prev != null) {
+            current.prev.next = newNode;
+        }
+        current.prev = newNode;
+        if (current == head) {
+            head = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw  new IndexOutOfBoundsException("Index is out of bound");
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw  new IndexOutOfBoundsException("Index is out of bound");
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw  new IndexOutOfBoundsException("Index is out of bound");
+        }
+        Node<T> current = head;
+        if (index == 0) {
+            T removeValue = head.value;
+            head = head.next;
+            if (head != null) {
+                head.prev = null;
+            } else {
+                tail = null;
+            }
+            size--;
+            return removeValue;
+        }
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        T removedValue = current.value;
+        if (current.prev != null) {
+            current.prev.next = current.next;
+        }
+        if (current.next != null) {
+            current.next.prev = current.prev;
+        }
+
+        if (current == tail) {
+            tail = current.prev;
+        }
+        size--;
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
```

Comment:
Create a separate method `unlink(Node node)` to handle the removal of a node. This will help to avoid repeating the unlinking logic in `remove(T object)` and `remove(int index)`.